File: LeetCode/Medium/1219-path-with-maximum-gold/1219-path-with-maximum-gold.cpp
class Solution {
public:
    int rowLen;
    int colLen;
    
    int getMaximumGold(vector<vector<int>>& grid) {
        rowLen = grid.size() - 1;
        colLen = grid[0].size() - 1;
        vector<vector<bool>> visited(rowLen + 1, vector<bool>(colLen + 1));
        int ret = 0;
        
        for (int i = 0; i < rowLen + 1; ++i)
            for (int j = 0; j < colLen + 1; ++j)
                ret = max(ret, dfs(i,j,0, grid, visited));
        
        return ret;
    }
    
    int dfs(int row, int col, int curr, vector<vector<int>>& grid, vector<vector<bool>>& visited ){
        if (grid[row][col] == 0 || visited[row][col])
            return curr;

        curr += grid[row][col];
        int temp = curr;
        visited[row][col] = true;

        if (row > 0)
            temp = max(temp, dfs(row-1, col, curr, grid, visited));
        if (row < rowLen)
            temp = max(temp, dfs(row+1, col, curr, grid, visited));
        if (col > 0)
            temp = max(temp, dfs(row, col-1, curr, grid, visited));
        if (col < colLen)
            temp = max(temp, dfs(row, col+1, curr, grid, visited));

        curr = temp;

        visited[row][col] = false;

        return curr;
    }
};

File: LeetCode/Medium/1219-path-with-maximum-gold/1219-path-with-maximum-gold.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        row_len = len(grid) - 1
        col_len = len(grid[0]) - 1
        visited = [[False] * (col_len + 1) for i in range(row_len + 1)]
        ret = 0
        
        def dfs(row, col, curr):
            if grid[row][col] == 0 or visited[row][col]:
                return curr
            
            curr += grid[row][col]
            temp = curr
            visited[row][col] = True
            
            if row > 0:
                temp = max(temp, dfs(row-1, col, curr))
            if row < row_len:
                temp = max(temp, dfs(row+1, col, curr))
            if col > 0:
                temp = max(temp, dfs(row, col-1, curr))
            if col < col_len:
                temp = max(temp, dfs(row, col+1, curr))
            
            curr = temp

            visited[row][col] = False
            
            return curr
        
        for i in range(row_len+1):
            for j in range(col_len+1):
                ret = max(ret, dfs(i,j,0))
        
        return ret